File: app/src/main/java/org/ligi/blexplorer/ServiceViewHolder.java
package org.ligi.blexplorer;

import android.bluetooth.BluetoothGattService;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;
import butterknife.ButterKnife;
import butterknife.InjectView;
import java.io.IOException;
import org.json.JSONException;
import org.json.JSONObject;
import org.ligi.axt.AXT;

class ServiceViewHolder extends RecyclerView.ViewHolder {

    @InjectView(R.id.name)
    TextView name;

    @InjectView(R.id.uuid)
    TextView uuid;

    @InjectView(R.id.type)
    TextView type;

    public ServiceViewHolder(final View itemView) {
        super(itemView);
        ButterKnife.inject(this, itemView);
    }

    public void applyService(final BluetoothGattService service) {
        uuid.setText(service.getUuid().toString());
        type.setText(DevicePropertiesDescriber.describeServiceType(service));

        final String serviceKey = service.getUuid().toString().split("-")[0];
        final String cleanServiceKey = serviceKey.replaceFirst("^0+(?!$)", ""); // remove leading zeroes

        try {
            final JSONObject jsonObject = new JSONObject(AXT.at(name.getContext().getAssets().open("services.json")).readToString());
            name.setText(jsonObject.getJSONObject(cleanServiceKey).getString("name"));
        } catch (IOException | JSONException e) {
            name.setText("unknown");
        }
    }
}

File: android/src/main/java/org/ligi/blexplorer/services/ServiceViewHolder.kt
package org.ligi.blexplorer.services

import android.bluetooth.BluetoothGattService
import android.support.v7.widget.RecyclerView
import org.ligi.blexplorer.App
import org.ligi.blexplorer.characteristics.CharacteristicActivity
import org.ligi.blexplorer.databinding.ItemServiceBinding
import org.ligi.blexplorer.util.DevicePropertiesDescriber
import org.ligi.kaxt.startActivityFromClass

class ServiceViewHolder(private val binding: ItemServiceBinding) : RecyclerView.ViewHolder(binding.root) {

    fun applyService(service: BluetoothGattService) {
        itemView.setOnClickListener { v ->
            App.service = service
            v.context.startActivityFromClass(CharacteristicActivity::class.java)
        }
        binding.uuid.text = service.uuid.toString()
        binding.type.text = DevicePropertiesDescriber.describeServiceType(service)
        binding.name.text = DevicePropertiesDescriber.getServiceName(itemView.context, service, "unknown")
    }
}
